File: src/abstraction/v1-8-9/java/net/eduard/api/lib/abstraction/Blocks_v1_8_R3.java
package net.eduard.api.lib.abstraction;

import net.minecraft.server.v1_8_R3.*;
import org.bukkit.Location;
import org.bukkit.craftbukkit.v1_8_R3.CraftWorld;
import org.bukkit.craftbukkit.v1_8_R3.block.CraftBlock;
import org.jetbrains.annotations.NotNull;

public class Blocks_v1_8_R3 extends CraftBlock implements Blocks {


    private transient final WorldServer worldServer;

    public WorldServer getWorldServer(){return worldServer;}

    public Blocks_v1_8_R3(int x, int y, int z, WorldServer worldServer) {
        super(null, x, y, z);
        this.worldServer = worldServer;
    }
    public Blocks_v1_8_R3(Location location) {
        this(location.getBlock());
    }

    public Blocks_v1_8_R3(org.bukkit.block.Block block) {
        this(block.getX(), block.getY(), block.getZ(), ((CraftWorld) block.getWorld()).getHandle());
    }

    @Override
    public org.bukkit.World getWorld() {
        return worldServer.getWorld();
    }

    @Override
    public org.bukkit.Chunk getChunk() {
        return worldServer.getChunkAt(getX() >> 4, getZ() >> 4).bukkitChunk;
    }

    @Override
    public int getTypeId() {
        final ChunkSection section = getSection();

        final IBlockData type = section.getType(getX() & 0xF, getY() & 0xF, getZ() & 0xF);
        final Block block = type.getBlock();

        return Block.getId(block);
    }

    @Override
    public byte getData() {
        final ChunkSection section = getSection();

        final IBlockData type = section.getType(getX() & 0xF, getY() & 0xF, getZ() & 0xF);
        final Block block = type.getBlock();

        return (byte) block.toLegacyData(type);
    }

    @Override
    public boolean setTypeIdAndData(int type, byte data, boolean applyPhysics) {
        final ChunkSection section = getSection();

        final int x = getX() & 0xF;
        final int y = getY() & 0xF;
        final int z = getZ() & 0xF;



        int combined = type + (data << 12);
        IBlockData blockData = net.minecraft.server.v1_8_R3.Block
                .getByCombinedId(combined);

        if (blockData == section.getType(x, y, z)) {
            return false;
        }

        section.setType(x, y, z, blockData);


        if (applyPhysics) {

            worldServer.notify(new BlockPosition(getX(), getY(), getZ()));
        }

        return true;
    }
    @Override
    public void setTypeAndData(@NotNull org.bukkit.Material material, int data) {
        setTypeIdAndData(material.getId(), (byte) data, true);
    }
    @Override
    public org.bukkit.block.Block getRelative(int modX, int modY, int modZ) {
        return new Blocks_v1_8_R3(getX() + modX, getY() + modY, getZ() + modZ, worldServer);
    }

    private ChunkSection getSection() {
        final Chunk chunk = worldServer.chunkProviderServer
                .originalGetChunkAt(getX() >> 4, getZ() >> 4);

        ChunkSection chunkSection = chunk.getSections()[getY() >> 4];
        if (chunkSection == null) {
            chunkSection
                    = chunk.getSections()[getY() >> 4]
                    = new ChunkSection(getY() >> 4 << 4, !worldServer.worldProvider.o());
        }

        return chunkSection;
    }


}

File: MineNMS-1_8_9/src/main/java/Blocks_v1_8_R3.kt
package net.eduard.api.lib.abstraction

import net.minecraft.server.v1_8_R3.*
import org.bukkit.Bukkit
import org.bukkit.Chunk
import org.bukkit.Material
import org.bukkit.World
import org.bukkit.block.Block
import org.bukkit.craftbukkit.v1_8_R3.CraftChunk
import org.bukkit.craftbukkit.v1_8_R3.block.CraftBlock
import org.bukkit.craftbukkit.v1_8_R3.CraftWorld

/**
 *
 */
class Blocks_v1_8_R3(
    x: Int, y: Int, z: Int, blockChunk: Chunk?,
    world: World
) : CraftBlock((blockChunk as CraftChunk?), x, y, z),
    Blocks {
    val position = BlockPosition(x, y, z)
    val worldServer: WorldServer = (world as CraftWorld).handle
    var chunkCache: net.minecraft.server.v1_8_R3.Chunk? =
        (blockChunk as CraftChunk?)?.handle ?: worldServer.chunkProviderServer.originalGetChunkAt(x shr 4, z shr 4)

    override fun getWorld(): World {
        return worldServer.world
    }


    override fun getChunk(): Chunk {
        return worldServer.getChunkAt(x shr 4, z shr 4).bukkitChunk
    }

    override fun getType(): Material {
        return Material.getMaterial(typeId)
    }

    override fun getTypeId(): Int {
        val section = section ?: return 0
        val type = section.getType(x and 0xF, y and 0xF, z and 0xF)
        val block = type.block
        return net.minecraft.server.v1_8_R3.Block.getId(block)
    }

    override fun getData(): Byte {
        val section = section ?: return 0
        val type = section.getType(x and 0xF, y and 0xF, z and 0xF)
        val block = type.block
        return block.toLegacyData(type).toByte()
    }

    override fun setType(material: Material) {
        setTypeIdAndData(material.id, 0, false)
    }

    override fun setTypeAndData(material: Material, data: Int): Boolean {
        return setTypeIdAndData(material.id, data.toByte(), false)
    }

    override fun setTypeIdAndData(type: Int, data: Byte, applyPhysics: Boolean): Boolean {
        val section = section ?: return false
        val xSec = x and 0xF
        val ySec = y and 0xF
        val zSec = z and 0xF
        val combined: Int = type + (data.toInt() shl 12)
        val blockData = net.minecraft.server.v1_8_R3.Block.getByCombinedId(combined)
        if (blockData === section.getType(xSec, ySec, zSec)) {
            //println("Travou aqui type $type  / $data")
            return false
        }
        section.setType(xSec, ySec, zSec, blockData)
        //worldServer.x(position)
        // altera a luz emitida do bloco (eles ficam igual a glowstone)
        //section.b(x, y, z, 0)
        //altera a luz do ceu aparentemente setar 15 n ta ajudando
        //section.a(xSec, (y shr 0xF) and 0xF, zSec, 15)

        if (applyPhysics) {
            // worldServer.notifyAndUpdatePhysics(position)
        } else {
            //println("notificando mudanca")

        }
        // Mine.broadcast("Modificando bloco")
        worldServer.x(position)
        Bukkit.getScheduler().runTask(Bukkit.getPluginManager().plugins.first(), this::sendPacket)
        // sendPacket()
        // Bukkit.getScheduler().runTask(Bukkit.getPluginManager().plugins.first(), this::sendPacket)
        //super.setTypeIdAndData(type, data, applyPhysics)
        return true
    }

    override fun sendPacket() {
        worldServer.notify(position)
    }

    override fun fixLightning() {
        //método responsavel por concertar as luzes porem da mais lag
        worldServer.x(position)
        //worldServer.getLightLevel(position)
        sendPacket()
    }


    override fun getRelative(modX: Int, modY: Int, modZ: Int): Block {
        return Blocks_v1_8_R3(x + modX, y + modY, z + modZ, chunkCache?.bukkitChunk, worldServer.world)
    }


    private val section: ChunkSection?
        get() {
            val x = x
            val y = y
            val z = z
            try {
                val chunk = chunkCache ?: return null
                // val chunk = worldServer.chunkProviderServer.originalGetChunkAt(x shr 4, z shr 4)
                val yBitShifted = (y shr 4) and 0xF
                if (yBitShifted >= 16) {
                    return null;
                }
                var chunkSection = chunk.sections[yBitShifted]
                if (chunkSection == null) {
                    chunkSection = ChunkSection(y shr 4 shl 4, !worldServer.worldProvider.o())
                    chunk.sections[y shr 4] = chunkSection
                    //chunkSection = chunk.sections[y shr 4]
                }
                return chunkSection
            } catch (ex: Exception) {
                return null;
            }
        }
}